plugins/usbdmx/JaRuleWidget.cpp

```diff
@@ -18,50 +18,253 @@
+    if (iface.num_altsetting != 1) {
+      continue;
+    }
+    const struct libusb_interface_descriptor &iface_descriptor =
+        iface.altsetting[0];
+    if (iface_descriptor.bInterfaceClass == LIBUSB_CLASS_VENDOR_SPEC &&
+        iface_descriptor.bInterfaceSubClass == 0xff &&
+        iface_descriptor.bInterfaceProtocol == 0xff) {
+      // Vendor class, subclass & protocol
+      for (uint8_t endpoint_index = 0; endpoint_index <
+           iface_descriptor.bNumEndpoints; endpoint_index++) {
+        const struct libusb_endpoint_descriptor &endpoint =
+            iface_descriptor.endpoint[endpoint_index];
+        if ((endpoint.bmAttributes & LIBUSB_TRANSFER_TYPE_MASK) !=
+             LIBUSB_TRANSFER_TYPE_BULK) {
+          continue;
+        }
+
+        uint8_t endpoint_address = endpoint.bEndpointAddress;
+        uint8_t endpoint_number = (
+            endpoint_address & LIBUSB_ENDPOINT_ADDRESS_MASK);
+        uint8_t endpoint_direction = (
+            endpoint_address & LIBUSB_ENDPOINT_DIR_MASK);
+
+        if (endpoint_direction == LIBUSB_ENDPOINT_IN) {
+          endpoint_map[endpoint_number].in_supported = true;
+          endpoint_map[endpoint_number].in_interface = iface_index;
+        }
+        if (endpoint_direction == LIBUSB_ENDPOINT_OUT) {
+          endpoint_map[endpoint_number].out_supported = true;
+          endpoint_map[endpoint_number].out_interface = iface_index;
+        }
+      }
+    }
+  }
+
+  m_adaptor->FreeConfigDescriptor(config);
+
+  if (!m_adaptor->OpenDevice(m_device, &m_usb_handle)) {
+    return false;
+  }
+
+  std::set<int> interfaces_to_claim;
+
+  EndpointMap::const_iterator endpoint_iter = endpoint_map.begin();
+  uint8_t port_index = 0;
+  for (; endpoint_iter != endpoint_map.end(); ++endpoint_iter) {
+    const EndpointCapabilties &capabilites = endpoint_iter->second;
+    if (capabilites.in_supported && capabilites.out_supported) {
+      interfaces_to_claim.insert(capabilites.in_interface);
+      interfaces_to_claim.insert(capabilites.out_interface);
+      OLA_INFO << "Found Ja Rule port at "
+               << static_cast<int>(endpoint_iter->first);
+      m_ports.push_back(new JaRuleWidgetPort(
+            m_executor, m_adaptor, m_usb_handle, endpoint_iter->first,
+            m_uid, port_index++));
+    }
+  }
+
+  std::set<int>::const_iterator iface_iter = interfaces_to_claim.begin();
+  for (; iface_iter != interfaces_to_claim.end(); ++iface_iter) {
+    if (m_adaptor->ClaimInterface(m_usb_handle, *iface_iter)) {
+      return false;
+    }
+  }
+
+  // Get the serial number (UID) of the device.
+  libusb_device_descriptor device_descriptor;
+  if (m_adaptor->GetDeviceDescriptor(m_device, &device_descriptor)) {
+    return false;
+  }
+
+  AsyncronousLibUsbAdaptor::DeviceInformation device_info;
+  if (!m_adaptor->GetDeviceInfo(m_device, device_descriptor, &device_info)) {
+    return false;
+  }
+
+  auto_ptr<UID> uid(UID::FromString(device_info.serial));
+  if (!uid.get()) {
```

Comment:
Do you mean the controller being the broadcast UID? We could add an option to skip the test as a command line flag.